#! /usr/bin/env/ Rscript
# See if can figure out if amount of compensation is surprising from proportions of trans effects, etc
# by Avery Davis Bell, begun 2025.01.09

# Get library location
if(length(.libPaths())==1){
  mylibloc <- .libPaths()[1]
}else{ # Presumes on PACE. This is not the best...
  mylibloc <- .libPaths()[grep(R.Version()$platform, .libPaths())]
}
cat(paste("Library location:", mylibloc, "\n"))

require(data.table, lib.loc = mylibloc)
require(argparser, lib.loc = mylibloc)
require(ggplot2, lib.loc = mylibloc)

# plotting theme
myggtheme<-theme_bw() +
  theme(axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 11),
        axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 11),
        legend.title = element_text(size = 12), legend.text = element_text(size=11),
        strip.text.x = element_text(size = 11), strip.text.y = element_text(size = 11), title = element_text(size = 14),
        strip.text.x.top = element_text(size = 11), strip.text.x.bottom = element_text(size = 11), strip.text.y.right = element_text(size = 11), strip.text.y.left = element_text(size = 11),
        plot.subtitle = element_text(size = 13))

#### Functions ####
# --- data functions
pandci<-function(x, n){
  # Output is one-row data.table with columns p, p_low95ci, p_high95ci
  # Deals with cases where n is 0
  myp<-x/n
  if(n>0){
    bres<-binom.test(x, n)
    out<-data.table(p = myp,  p_low95ci = bres$conf.int[1], p_high95ci = bres$conf.int[2])
  }else{
    out<-data.table(p = NaN,  p_low95ci = NaN, p_high95ci = NaN)
  }

  return(out)
}

transpropsone<-function(nregone, annotdt = NULL){
  # Gets the numbers, proportions of genes with trans effects with and without ASE, as well as the FET comparing these. For one strain's counts
  # In: nregone, one-row regulatory pattern counts. Columns used: conserved, trans, cis, cis-trans opposing, enhancing
  #     annotdt, optional one-row data.table to add to beginning of output
  # Out: one-row data.table with columns:
  #       <any in annotdt>
  #       n.trans.nonase, # trans-only regulated genes, obligatorily don't have ASE
  #       n.non.ase, # genes that don't have ASE but simplified to ones we feel strongly about (trans regulated or not ): conserved + trans-labeled genes
  #       n.cistransopposing, # cis-trans opposing (umbrella category is already in input)
  #       n.ase, # with any cis effect (those with ASE simplified to ones we feel strongly about, trans regulated or not): cis + enhancing + cis-trans opposing
  #       p.trans.ofnonase, n.trans.nonase/n.non.ase
  #       p.trans.ofnonase_low95, lower binomial 95% CI
  #       p.trans.ofnonase_hi95, upper binomial 95% CI
  #       p.comp.ofase, n.cistransopposing/n.ase
  #       p.comp.ofase_low95, lower binomial 95% CI
  #       p.com.pofase_hi95, upper binomial 95% CI
  #       or.fet, odds ratio of Fisher's Exact Test comparing these proportions
  #       pvalue.fet, p-value (two-sided) of Fisher's Exact Test comparing these proportions

  # Get numbers
  out<-data.table(n.trans.nonase = nregone$trans,
                  n.non.ase = nregone[, trans + conserved],
                  n.cistransopposing = nregone[, `cis-trans opposing`],
                  n.ase = nregone[, enhancing + cis + `cis-trans opposing`])

  # Get proportions
  ptnonase<-out[, pandci(n.trans.nonase, n.non.ase)]
  setnames(ptnonase, c("p.trans.ofnonase", "p.trans.ofnonase_low95", "p.trans.ofnonase_hi95"))
  pcase<-out[, pandci(n.cistransopposing, n.ase)]
  setnames(pcase, c("p.comp.ofase", "p.comp.ofase_low95", "p.comp.ofase_hi95"))

  # FET
  res<-fisher.test(matrix(unlist(out), nrow = 2))

  # Combine & return
  out<-data.table(annotdt, out, ptnonase, pcase,
                  or.fet = res$estimate,
                  pvalue.fet = res$p.value)
  return(out)
}

subonease<-function(ns, percent, annotdt = NULL){
  # Subtracts n genes from n.cistransopposing and n.ase, with n being percent% of the total  n.ase
  # Re-computes proportions, re-does FET
  # In: ns, one-row data.table with some of columns generated by transpropsone. Columns: n.trans.nonase, n.non.ase, n.cistransopposing, n.ase
  #     percent, percent (i.e., 10 = 10%/0.1) ASE genes to remove
  #     annotdt, optional one-row data.table to add to beginning of output
  # Out: one-row data.table with input columns (numbers changed for n.cistransopposing, n.ase - 0 if would've been negative) and:
  #         percentdown, what percentage n.ase were removed here
  #         number.rm, what NUMBER n.ase were removed here
  #         percentcomp, what percentage of n.cistransopposing were removed
  #         p.trans.ofnonase, same as input
  #         or.fet, Odds Ratio from Fisher's exact test on these (new) data
  #         pvalue.fet, p-value from Fisher's exact test on these (new) data
  #         p.comp.ofase, proportion from new data

  # Do N downsampling, get props
  nrm<-round(ns$n.ase*(percent/100))
  out<-data.table(n.trans.nonase = ns$n.trans.nonase,
                  n.non.ase = ns$n.non.ase,
                  n.cistransopposing = max(ns$n.cistransopposing - nrm, 0), # obligately keeps at 0
                  n.ase = ns$n.ase - nrm)
  out[, `:=`(p.trans.ofnonase = n.trans.nonase/n.non.ase,
             p.comp.ofase = n.cistransopposing/n.ase)]

  # FET
  res<-fisher.test(matrix(unlist(out[, 1:4, with =F]), nrow = 2))

  # Combine & return
  out<-data.table(annotdt,
                  percentdown = percent, number.rm = nrm, percentcomp = nrm/ns$n.cistransopposing*100,
                  out,
                  or.fet = res$estimate,
                  pvalue.fet = res$p.value)
  return(out)
}

subonede<-function(nregone, percent, annotdt = NULL){
  # Subtracts number based on percent given from actual
    ## subtract just from n.cistransopposing, keeping n.ase constant [basically, increasing # called cis]. Prop of OVERALL DE [trans + cis (+ enhancing)?] this would be

## keep thinking about this - simpler to just say would have had to miss DE when it was actually there as many times as called it, or similar
## (like, do based on actual numbers rather than this subtract a percent thing?)....Similar to the NEXT thing in script/notebook I'm thinking about

}

# --- Plotting functions
stackedbar<-function(pinhclass, mycolors, xcol = "strain", stackcol = "propName", stacknumcol = "prop", xorder = NA,
                     legendlabel = "", myxlab = "Strain", myylab = "Proportion genes", mytitle = "", mysubt = "",
                     myleglabels = c("", "")){
  # Modified from chrlocenrichment_asederpim.R; copied from ase_figure_functions.R
  # Makes stacked bar chart, one stacked bar per category in x col. Copied from other scripts.
  # In: pinhclass, data.table with all data to plot. Must have columns named with values of xcol, stackcol, stacknumcol
  #     mycolors, colors for bars: vector of length of categories in data; also used as ORDER for bars!! Names are categories, values are colors
  #     xcol, column containing category for X axis of bar chart
  #     stackcol, column containing category to split bars into
  #     stacknumcol, column containing numbers to plot as stacked bar
  #     xorder, OPTIONAL order in which to arrange categories on x axis (in xcol)
  #     legendlabel
  #     myxlab, myylab, mytitle, mysubt: plot labels as intuitively named
  #     myleglabels, provide legend label names (**in order they're plotted **) - in exorder!
  # Out: ggplot2 barplot object. Can facet this externally!

  # Format data
  pdata<-copy(pinhclass)
  pdata[,mystackcol:=factor(pdata[,get(stackcol)], levels = names(mycolors))] # relevel factor
  if(!is.na(xorder[1])){
    # Re-level factor
    pdata[,xcol:=factor(pdata[,get(xcol)], levels = xorder)]
  }else{pdata[,xcol:=get(xcol)]}

  plt<-ggplot(pdata, aes(xcol, eval(as.name(stacknumcol)))) +
    geom_bar(aes(fill = mystackcol), stat = "identity", position = "stack") + labs(fill = legendlabel) +
    xlab(myxlab) + ylab(myylab) +
    scale_fill_manual(values = mycolors, labels = myleglabels) + # provided colors
    theme_bw() + myggtheme

  return(plt)
}

#### Arguments & inputs ####
# --- Command line arguments
p<-arg_parser("See if can figure out if amount of compensation is surprising from proportions of trans effects, etc",
              name = "transpropcompensationtesting.R ", hide.opts = TRUE)

# Organizational & data input arguments
p<-add_argument(p, "--regpatns",
                help = "Counts of regulatory patterns within strain and geneset - *numbers_regpatterns.txt output of ase_de_cistransclassifications.R.
                **MUST be for only one 'CI' column value**",
                type = "character")
p<-add_argument(p, "--baseoutname",
                help = "Base name for all output files; should likely include strain, for example",
                type = "character",
                default = "out")
p<-add_argument(p, "--varsvsref",
                help = "Path to file containing columns strain (one row for each strain in main input), nvars (number variants vs. reference genome) [other columns optional];
                for ordering strains etc",
                type = "character")
p<-add_argument(p, "--outdir",
                help = "[[output-related]] Output directory. **NB: if you provide getwd() here (quote wrapped), current directory will be used",
                type = "character",
                default = "out")


# Parse arguments
cat("....Parsing arguments....\n")
p<-parse_args(p)

# Output directory
if(p$outdir=="getwd()"){
  p$outdir<-getwd()
}
if(!dir.exists(p$outdir)){dir.create(p$outdir, recursive = T)}
setwd(p$outdir)

# Read in data
straininfo<-fread(p$varsvsref, header = T)[order(nvars), ]
regns<-fread(p$regpatns, header = T)

#### Proportion genes with trans effects vs. ASE genes with trans-compensatory effects ####
# --- compute proportions, results (5plusUnqAlns & 5plusUnqAlns_exclhypdivbadcov)
props<-rbindlist(lapply(c("5plusUnqAlns", "5plusUnqAlns_exclhypdivbadcov"), function(g){
  rbindlist(lapply(straininfo$strain, function(strn){
    transpropsone(nregone = regns[strain==strn & geneset==g, ], annotdt = data.table(strain = strn, geneset = g))
  }))
}))
# Save
write.table(props, file.path(p$outdir, paste0(p$baseoutname, "_transprops_statres_ofNotASEandASE.txt")),
            sep = "\t", quote = F, row.names = F)

# --- Figure out how to plot this nicely
## long format for bar plotting
props.long<-rbind(props[, .(strain, geneset, denomdescrip = "No ASE/cis", comp.trans = T, # F is 1 - this
                            n = n.trans.nonase, n.tot = n.non.ase,
                                       p = p.trans.ofnonase, p.low95 = p.trans.ofnonase_low95, p.hi95 = p.trans.ofnonase_hi95,
                                       or.fet, pvalue.fet)],
                  props[, .(strain, geneset, denomdescrip = "ASE/cis", comp.trans = T,
                            n = n.cistransopposing, n.tot = n.ase,
                            p = p.comp.ofase, p.low95 =  p.comp.ofase_low95, p.hi95 =  p.comp.ofase_hi95,
                            or.fet, pvalue.fet)],
                   props[, .(strain, geneset, denomdescrip = "No ASE/cis", comp.trans = F,
                            n = n.non.ase - n.trans.nonase, n.tot = n.non.ase,
                            p = 1 - p.trans.ofnonase, p.low95 = p.trans.ofnonase_low95, p.hi95 = p.trans.ofnonase_hi95,
                            or.fet, pvalue.fet)],
                  props[, .(strain, geneset, denomdescrip = "ASE/cis", comp.trans = F,
                            n = n.ase - n.cistransopposing, n.tot = n.ase,
                            p = 1 - p.comp.ofase, p.low95 =  p.comp.ofase_low95, p.hi95 =  p.comp.ofase_hi95,
                            or.fet, pvalue.fet)]) ## 95 CIs probably wrong here (for comp.trans = F)

# save in case need for other plotting
write.table(props.long, file.path(p$outdir, paste0(p$baseoutname, "_transprops_statres_longformat_ofNotASEandASE.txt")),
            sep = "\t", quote = F, row.names = F)

#--- Make bar plot
twocols<-c("gray", "black")
names(twocols)<-c(FALSE, TRUE)

props.long[, strain:=factor(strain, levels = straininfo[order(nvars, strain), strain])]

propbar<-stackedbar(pinhclass = props.long,
                            mycolors = twocols, xcol = "denomdescrip", stackcol = "comp.trans",
                            stacknumcol = "p", xorder = c("No ASE/cis", "ASE/cis"), legendlabel = "",
                            myxlab = "Genes", myylab = "Proportion genes", myleglabels = c("No trans/compensated", "Trans/compensated")) +
  facet_grid(geneset~strain) + theme(legend.position = "bottom")
      # for manuscript, would add FET test results; possibly collapse to just be one panel rather than per-strain
pdf(file.path(p$outdir, paste0(p$baseoutname, "_transprops_ofNotASEandASE_barplot.pdf")),
    9, 5.5)
print(propbar)
invisible(dev.off())


# --- Connected pointrange plot (everyone in same panel)
props.long[, denomdescrip:=factor(denomdescrip, levels = c("No ASE/cis", "ASE/cis"))]
propdot<-ggplot(data = props.long[comp.trans==T], aes(denomdescrip, p)) +
  geom_line(aes(group = strain, color = strain), alpha = 0.4) +
  geom_pointrange(aes(ymin = p.low95, ymax = p.hi95, color = strain), alpha = 0.4) +
  xlab("Genes") + ylab("Proportion with trans effect\n(as DE or compensation, respectively)") +
  facet_wrap("geneset") +
  myggtheme

pdf(file.path(p$outdir, paste0(p$baseoutname, "_transprops_ofNotASEandASE_dotplot.pdf")),
    8, 6)
print(propdot)
invisible(dev.off())


#### Comparisons to show this probably isn't just messing up detecting DE/over-calling ASE ####
# --- Stats/numbers: How many ASE calls would have to be spurious for proportions to even out? how many DE calls would have to be missed for proportions to even out?
# How many ASE calls would have to be spurious for proportions to even out?
## Subtract same n% from n.cistransopposing and n.ase, re-compute p, re-do FET
## Get data
downase<-rbindlist(lapply(seq(1, 60), function(p){
  rbindlist(lapply(c("5plusUnqAlns", "5plusUnqAlns_exclhypdivbadcov"), function(g){
  rbindlist(lapply(straininfo$strain, function(strn){
    subonease(ns = props[strain==strn & geneset==g,],
              percent = p, annotdt = data.table(strain = strn, geneset = g))
  }))
}))
}))

## Plot/Analyze
## *****take direction into consideration, FET p-values are 2 sided
downase[, strain:=factor(strain, levels = straininfo[order(nvars, strain), strain])]
plt.downase.together<-ggplot(data = downase, aes(percentdown, -1*log10(pvalue.fet))) +
  geom_line(aes(color = strain)) +
  geom_point(aes(color = strain, shape = p.comp.ofase>p.trans.ofnonase)) +
  labs(shape = "Compensated proportion of ASE >\n trans proportion of non-ASE") +
  xlab("Percent ASE genes REMOVED") + ylab("-log10(FET p-value) (higher = more significant)") +
  facet_wrap("geneset") + myggtheme + theme(legend.position = "bottom") +
  geom_hline(yintercept = -1*log10(0.05/7), lty = "dashed") # bonferroni for N strains

plt.downase.sep<-ggplot(data = downase, aes(percentdown, -1*log10(pvalue.fet))) +
  geom_line(aes(color = strain)) +
  geom_point(aes(color = strain, shape = p.comp.ofase>p.trans.ofnonase)) +
  labs(shape = "Compensated proportion of ASE >\n trans proportion of non-ASE") +
  xlab("Percent ASE genes REMOVED") + ylab("-log10(FET p-value) (higher = more significant)") +
  facet_grid(geneset~strain, scales = "free") + myggtheme + theme(legend.position = "bottom") +
  geom_hline(yintercept = -1*log10(0.05/7), lty = "dashed") # bonferroni for N strains

## Save data, plots - own directory
dasedir<-file.path(p$outdir, "reduceaseretest")
if(!dir.exists(dasedir)){dir.create(dasedir, recursive = T)}

write.table(downase, file.path(dasedir, paste0(p$baseoutname, "_downase_data.txt")),
            sep = "\t", quote = F, row.names = F)
pdf(file.path(dasedir, paste0(p$baseoutname, "_downase_premovedVsFETpval.pdf")),
    10, 8)
print(plt.downase.together)
print(plt.downase.sep)
invisible(dev.off())

# IF ASE calls are right, would have to miss DE at (XX%) of genes we see it at
## DE here: cis + enhancing + trans + compensating + overcompensating (trying to be conservative)
## DE 'missed' if compensated ones are really cis: compensatory
dens<-regns[geneset%in%c("5plusUnqAlns", "5plusUnqAlns_exclhypdivbadcov"),
            .(strain, geneset, DECalled = (cis+enhancing+trans + compensating + overcompensating),
                DEMissed = compensatory)]
dens[, `:=`(p.ofcalled = DEMissed/DECalled,
            p.ofcalledPlusMissed = DEMissed/(DEMissed + DECalled))]

dens[, strain:=factor(strain, levels = straininfo[order(nvars, strain), strain])]
dens<-dens[order(strain), ]

write.table(dens, file.path(p$outdir, paste0(p$baseoutname, "_demissedInASE_ofNDE.txt")),
            sep = "\t", quote = F, row.names = F)

#### Session info ####
cat("....transpropcompensationtesting.R processing complete! Session information:....\n")
sessionInfo()
